Update React page (#593)
rescript-association/rescript-lang.org: pages/docs/react/latest/migrate-from-v3.mdx

---
title:
description: "Migrate from JSX v3"
canonical: "/docs/react/latest/jsx-v3-v4"
---

# Migrate from JSX v3

> JSX v4 introduces a new idiomatic record-based representation of components which is incompatible with v3. Because of this, either the entire project or dependencies need to be compiled in V4 mode, or some compatibility features need to be used to mix V3 and V4 in the same project. This page describes how to migrate from v3 to v4.

Reviewer: ```suggestion
JSX v4 introduces a new idiomatic record-based representation of component props which is incompatible with v3. Because of this, either the entire project or dependencies need to be compiled in V4 mode, or some compatibility features need to be used to mix V3 and V4 in the same project. This page describes how to migrate from v3 to v4.
```

## Configuration

Remove the existing JSX configuration from `bsconfig.json`:

```json
{
  "reason": { "react-jsx": 3 }
}
```

Then add the new JSX configuration:

```json
{
  "jsx": { "version": 4 }
}
```

**Note** JSX v4 requires the rescript compiler 10.1 or higher, and rescript-react version 0.11 or higher. In addition, react version 18.0 is required.

### Classic and Automatic Mode

Classic mode is the default and generates calls to `React.createElement` just as with V3.

```json
{
  "jsx": { "version": 4, "mode": "classic" }
}
```

Automatic mode is an experimental mode that generate calls to `_jsx` functions (similar to TypeScript's `react-jsx` mode)

```json
{

>   "jsx": { "version": 4, "mode": "Automatic" }

Reviewer 2: lowercase: `automatic`

Author: Thanks! fixed https://github.com/rescript-association/rescript-lang.org/pull/593/commits/9bf0ed5bd89873475405bc9b7004a0385a82f13b

>   "jsx": { "version": 4, "mode": "Automatic" }

Author: ```suggestion
  "jsx": { "version": 4, "mode": "automatic" }
```

}
```

> ### File-leve config

Reviewer: ```suggestion
### File-level config
```

The top-level attribute `@@jsxConfig` is used to update the `jsx` config for the rest of the file (or until the next config update). Only the values mentioned are updated, the others are left unchanged.

```res
@@jsxConfig({ version: 4, mode: "automatic" })

module Wrapper = {
  module R1 = {
    @react.component  // V4 and new _jsx transform
    let make = () => body
  }

  @@jsxConfig({ version: 4, mode: "classic" })

  module R2 = {
    @react.component  // V4 with `React.createElement`
    let make = () => body
  }
}

@@jsxConfig({ version: 3 })

@react.component  // V3
let make = () => body
```

### v3 compatible mode

JSX v3 is still available with the latest version of compiler and rescript-react.

```json
{
  "jsx": { "version": 3, "v3-dependencies": ["rescript-relay"] },
  "bsc-flags": ["-open ReactV3"]
}
```

> To build certain dependencies in V3 compatibility mode, whatever the version used in the root project, use `"v3-dependencies"`: the listed dependencies will be built in V3 mode, and in addition `-open ReatcV3` is added to the compiler options.

Reviewer: ```suggestion
To build certain dependencies in V3 compatibility mode, whatever the version used in the root project, use `"v3-dependencies"`. The listed dependencies will be built in V3 mode, and in addition `-open ReactV3` is added to the compiler options.
```

## Migration of V3 components

Some components in existing projects are written in a way that is dependent on the v3 internal representation. Here are a few examples of how to convert them to v4.

### `makeProps` does not exist in v4

Rewrite this:

```res
// V3
module M = {
  @obj external makeProps: (~msg: 'msg, ~key: string=?, unit) => {"msg": 'msg} = ""

  let make = (~msg) => <div> {React.string(msg)} </div>
}
```

To this:

```res
// V4
module M = {
  type props<'msg> = {msg: 'msg}
  let make = props => <div> {React.string(props.msg)} </div>
}
```

### React.Context

Rewrite this:

```res
module Context = {
  let context = React.createContext(() => ())

  module Provider = {
    let provider = React.Context.provider(context)

    @react.component
    let make = (~value, ~children) => {
      React.createElement(provider, {"value": value, "children": children}) // Error
    }
  }
}
```

To this:

```res
module Context = {
  let context = React.createContext(() => ())

  module Provider = {

>     let make = React.Context.provider(context)

Reviewer: `type props` is missing

  }
}
```

> ### React.forwardRef(Discouraged)

Reviewer: ```suggestion
### React.forwardRef (Discouraged)
```
